lib/asciidoctor/syntax_highlighter.rb

```diff
@@ -0,0 +1,200 @@
+    # opts   - A Hash of options that control syntax highlighting:
+    #          :nowrap - a Boolean that indicates that wrapping should be disabled
+    #
+    # Returns the highlighted source String wrapped in preformatted tags (e.g., pre and code)
+    def format node, lang, opts
+      raise ::NotImplementedError, 'Asciidoctor::SyntaxHighlighter must implement the format method'
+    end
+
+    # Public: Indicates whether this syntax highlighter wants to write a stylesheet to disk.
+    #
+    # This method is only called if both the linkcss and copycss attributes are set on the document.
+    #
+    # doc  - the Document in which this syntax highlighter is being used
+    #
+    # Returns a Boolean indicating whether the write_stylesheet method should be called.
+    def write_stylesheet? doc; end
+
+    # Public: Writes the stylesheet to support the highlighted source(s) to disk.
+    #
+    # doc  - the Document in which this syntax highlighter is being used
+    # to_dir - the absolute String path of the stylesheet output directory.
+    #
+    # Returns nothing.
+    def write_stylesheet doc, to_dir
+      raise ::NotImplementedError, 'Asciidoctor::SyntaxHighlighter must implement the write_stylesheet method if the write_stylesheet? method returns true'
+    end
+
+    private_class_method def self.included into
+      into.extend Config
+    end
+
+    module Config
+      # Public: Statically register the current class in the registry for the specified names.
+      #
+      # Returns nothing.
+      private def register_for *names
+        SyntaxHighlighter.register self, *names
+      end
+    end
+
+    module Factory
+      # Public: Associate the syntax highlighter class or object with the specified names.
+      #
+      # Returns nothing.
+      def register syntax_highlighter, *names
+        names.each {|name| registry[name] = syntax_highlighter } && nil
+      end
+
+      # Public: Retrieve the syntax highlighter class or object registered for the specified name.
+      #
+      # name - the String name of the syntax highlighter to retrieve.
+      #
+      # Returns the SyntaxHighlighter class or instance registered for this name.
+      def for name
+        registry[name]
+      end
+
+      # Public: Resolve the name to a syntax highlighter instance, if found in the registry.
+      #
+      # name - the String name of the syntax highlighter to create
+      # doc  - the Document in which this syntax highlighter is being used
+      #
+      # Returns a SyntaxHighlighter instance for the specified name.
+      def create name, doc
+        if (syntax_hl = self.for name)
+          syntax_hl = syntax_hl.new name, doc if ::Class === syntax_hl
+          raise ::NameError, %(#{syntax_hl.class.name} must specify a value for `name') unless syntax_hl.name
+          syntax_hl
+        end
+      end
+
+      private def registry
+        @registry ||= {}
+      end
+    end
+
+    class CustomFactory
```

Comment:
We should probably add a test with a `CustomFactory` (or a `Factory`)

Reply:
Good point. I'll add one.

As a quick question, I'm thinking the CustomFactory would be more useful if defined as:

```rb
class CustomFactory
  include Factory

  def initialize registry = nil 
    @registry = registry
  end
end
```

So now you can do:

```rb
factory = Asciidoctor::SyntaxHighlighter::CustomFactory.new name => clazz
```

Then we'd define the DefaultFactoryProxy (which allows the custom factory to fall back to the built-in defaults), as follows:

```rb
class DefaultFactoryProxy < CustomFactory
  include DefaultFactory

  def for name
    (@registry.key? name) ? @registry[name] : super
  end
end
```

wdyt?

Reply:
> As a quick question, I'm thinking the CustomFactory would be more useful if defined as

Why not, as far as I understand it's a shortcut:

```rb
syntax_hl_factory = Asciidoctor::SyntaxHighlighter::CustomFactory.new
syntax_hl_factory.register clazz, name

# shortcut
syntax_hl_factory = Asciidoctor::SyntaxHighlighter::CustomFactory.new name => clazz
```

> Then we'd define the DefaultFactoryProxy (which allows the custom factory to fall back to the built-in defaults), as follows:

But we could define a custom strategy to resolve the syntax highlighter in the CustomFactory ?
So the `DefaultFactoryProxy` should first check that the `CustomFactory.for name` does not return something and then fallback to the built-in no ? Or maybe I'm reading backward...

Reply:
> So the DefaultFactoryProxy should first check that the CustomFactory.for name does not return something and then fallback to the built-in no ?

That's exactly what it does. But it also allows the entry to be nil (which is why we check the key).

Reply:
`DefaultFactory` defines a `for` method and `DefaultFactoryProxy` includes `DefaultFactory` but overrides the method so `super` will call `DefaultFactory.for` right ?
But now you also extend `CustomFactory` which can define a `for` method too ? And now I'm lost :laughing:

Reply:
The include supersedes the extend. So `super` is DefaultFactory.for. All the extend does is give us the constructor (and the type). (In the case of Opal, `super` will jump to Factory.for, so everything still works).

Reply:
If I execute the code it will make things easier to understand but on screen I'm like 🤯